imap_processing/ialirt/l0/parse_mag.py

```diff
@@ -0,0 +1,246 @@
+"""Functions to support I-ALiRT MAG packet parsing."""
+
+import logging
+
+import numpy as np
+import xarray as xr
+
+from imap_processing.ialirt.l0.mag_l0_ialirt_data import (
+    Packet0,
+    Packet1,
+    Packet2,
+    Packet3,
+)
+from imap_processing.ialirt.utils.grouping import find_groups
+from imap_processing.ialirt.utils.time import calculate_time
+
+logger = logging.getLogger(__name__)
+
+
+def get_pkt_counter(status_values: xr.DataArray) -> xr.DataArray:
+    """
+    Get the packet counters.
+
+    Parameters
+    ----------
+    status_values : xr.DataArray
+        Status data.
+
+    Returns
+    -------
+    pkt_counters : xr.DataArray
+        Packet counters.
+    """
+    # mag_status is a 24 bit unsigned field
+    # The leading 2 bits of STATUS are a 2 bit 0-3 counter
+    pkt_counter = (status_values >> 22) & 0x03
+
+    return pkt_counter
+
+
+def get_status_data(status_values: xr.DataArray, pkt_counters: xr.DataArray) -> dict:
+    """
+    Get the status data.
+
+    Parameters
+    ----------
+    status_values : xr.DataArray
+        Status data.
+    pkt_counters : xr.DataArray
+        Packet counters.
+
+    Returns
+    -------
+    combined_packets : dict
+        Decoded packets.
+    """
+    decoders = {
+        0: Packet0,
+        1: Packet1,
+        2: Packet2,
+        3: Packet3,
+    }
+
+    combined_packets = {}
+
+    for pkt_num, decoder in decoders.items():
+        status_subset = status_values[pkt_counters == pkt_num]
+        decoded_packet = decoder(int(status_subset))
+        combined_packets.update(vars(decoded_packet))
+
+    return combined_packets
+
+
+def get_bytes(val: int) -> list[int]:
+    """
+    Extract three bytes from a 24-bit integer.
+
+    Parameters
+    ----------
+    val : int
+        24-bit integer value.
+
+    Returns
+    -------
+    list[int]
+        List of three extracted bytes.
+    """
+    return [
+        (val >> 16) & 0xFF,  # Most significant byte (Byte2)
+        (val >> 8) & 0xFF,  # Middle byte (Byte1)
+        (val >> 0) & 0xFF,  # Least significant byte (Byte0)
+    ]
+
+
+def extract_magnetic_vectors(science_values: xr.DataArray) -> dict:
+    """
+    Extract the magnetic vectors.
+
+    Parameters
+    ----------
+    science_values : xr.DataArray
+        Science data.
+
+    Returns
+    -------
+    vectors : dict
+        Magnetic vectors.
+    """
+    # Convert each 24-bit value to its three constituent bytes
+    science0 = get_bytes(int(science_values[0]))
+    science1 = get_bytes(int(science_values[1]))
+    science2 = get_bytes(int(science_values[2]))
+    science3 = get_bytes(int(science_values[3]))
+
+    # Primary sensor:
+    pri_x = (science0[0] << 8) | science0[1]
+    pri_y = (science0[2] << 8) | science1[0]
```

Comment:
I find the `get_bytes()` above to be more distracting / confusing to follow than direct bit manipulation inline here. Could this be something like the following (probably wrong, but hopefully it gets the idea across)?

```
pri_x = science_values[0] >> 8
pri_y = ((science_values[0] << 16) & 0xFF00) | (science_values[1] >> 8)
```